Microsoft.Research/Contracts/System.Core.Contracts/System.Linq.Expressions.DynamicExpression.cs

```diff
@@ -49,34 +51,39 @@ protected internal override Expression Accept (ExpressionVisitor visitor)
       return default(Expression);
     }
 
+    [Pure]
     public DynamicExpression Update (IEnumerable<Expression> arguments)
     {
+      Contract.Ensures(Contract.Result<DynamicExpression>() != null);
```

Comment:
> It appears that as a general rule they promote nulls to empty collections, unless that collection is expected to be non-empty. If that's the case they require it be non-null.

:arrow_right: I think this is the answer. For construction and transformation, the library treats null collections as empty collections, which means null is valid anywhere that empty is. Since the fastest path through `ToReadOnly` for the case of an empty collection is when null is provided, it's safe to assume client code leverages this, especially in performance-sensitive code.

:thought_balloon: In a world where System.Collections.Immutable preceded System.Linq.Expressions, it's likely that all of these parameters would be `ImmutableArray<T>` and would require non-default instances to express empty. But we don't live in that world. :smile: